usr.man/man3/crypt.3
.\" Copyright (c) 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)crypt.3	8.2 (Berkeley) 12/11/93
.\"
.TH CRYPT 3 "28 January 1997" GNO "Library Routines"
.SH NAME
.BR crypt ,
.BR setkey ,
.BR encrypt ,
\- simple encryption
.SH SYNOPSIS
#include <unistd.h>
.sp 1
char
*\fBcrypt\fR (const char *\fIkey\fR, const char *\fIsalt\fR);
.br
int
\fBsetkey\fR (char *\fIkey\fR);
.br
int
\fBencrypt\fR (char *\fIblock\fR, int \fIflag\fR);
.SH DESCRIPTION
The
.BR crypt 
function
performs password encryption.
It is derived from an algorithm given in 
.I "Computer Networks"
by Andy Tanenbaum.
.LP
The first argument to
.BR crypt
.RI ( key )
is a
.BR NULL -terminated
string (normally a password typed by a user).
The second,
.IR salt ,
is a two-character array that should be randomly generated by the caller
in the case of encrypting a new password, or should be taken as the
first two characters of the 
.BR /etc/passwd
password entry in the case of validating a password.
.LP
The
functions,
.BR encrypt
and
.BR setkey 
allow limited access to the encryption
algorithm itself.
The
.IR key
argument to
.BR setkey 
is a 64 character array of
binary values (numeric 0 or 1).
A 56-bit key is derived from this array by dividing the array
into groups of 8 and ignoring the last bit in each group.
.LP
The
.BR encrypt 
argument
.I block
is also a 64 character array of
binary values.
If the value of
.I flag
is 0,
the argument
.I block
is encrypted, otherwise it
is decrypted.
The encryption or decryption is returned in the original
array
.I block
after using the
key specified
by
.BR setkey 
to process it.
.LP
The function
.BR crypt 
returns a pointer to the encrypted value (which is formatted as 
printable ASCII characters) on success and NULL on failure.
The functions
.BR setkey 
and
.BR encrypt 
return 0 on success and 1 on failure.
.LP
Use of these routines requires linking with the 
.BR libcrypt
library.
.SH SEE ALSO
.BR login (1),
.BR passwd (1),
.BR getpass (3),
.BR passwd (5)
.sp
.RS
.I "Mathematical Cryptology for Computer Scientists and Mathematicians"
by
Wayne Patterson.
1987.
Volume ISBN 0-8476-7438-X.
.RE
.sp 1
.RS
.I "Password Security: A Case History"
by
R. Morris,
Ken Thompson.
.IR "Communications of the ACM" ,
vol. 22, pp. 594-597.  November 1979.
.RE
.sp 1
.RS
.I "DES will be Totally Insecure within Ten Years"
by
M.E. Hellman.
.IR "IEEE Spectrum" ,
vol. 16, pp. 32-39.  July 1979.
.RE
.SH BUGS
The
.BR crypt 
function leaves its result in an internal static object and returns
a pointer to that object.
Subsequent calls to
.BR crypt 
will modify the same object.


lib/libcrypt/crypt.c
/*
 * From Andy Tanenbaum's book "Computer Networks", rewritten in C.
 *
 * Many performance modifications made
 * January 19-22, 1992 by Jawaid Bazyar
 * Copyright 1992-1997, Procyon Inc.
 *
 * $Id: crypt.c,v 1.1 1998/02/09 08:25:33 taubert Exp $
 *
 * This file is formatted with tab stops every 8 columns.
 */

#ifdef __ORCAC__
/* segment "libc_gen__"; *//* don't segment this until crypt.asm is done */
#endif

#include <stdlib.h>
#include <string.h>

/* These symbols have been moved to or are referenced from crypta.asm */
#ifndef NOTDEFINED
#define f(i,key,a,x)		__crypt_f(i, key, a, x)
#define transpose(data,t,n)	__crypt_transpose(data, t, n)
#define rotate(key)		__crypt_rotate(key)
#define s_boxes			__crypt_s_boxes
#define rots			__crypt_rots
#define EP			__crypt_EP
#endif

struct block {
	unsigned char b_data[64];
};

struct ordering {
	unsigned char o_data[64];
};

static struct block key;	/* gdr added "static" */

static struct ordering InitialTr = {
	58,50,42,34,26,18,10, 2,60,52,44,36,28,20,12, 4,
	62,54,46,38,30,22,14, 6,64,56,48,40,32,24,16, 8,
	57,49,41,33,25,17, 9, 1,59,51,43,35,27,19,11, 3,
	61,53,45,37,29,21,13, 5,63,55,47,39,31,23,15, 7,
};

static struct ordering FinalTr = {
	40, 8,48,16,56,24,64,32,39, 7,47,15,55,23,63,31,
	38, 6,46,14,54,22,62,30,37, 5,45,13,53,21,61,29,
	36, 4,44,12,52,20,60,28,35, 3,43,11,51,19,59,27,
	34, 2,42,10,50,18,58,26,33, 1,41, 9,49,17,57,25,
};

static struct ordering swap = {
	33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,
	49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,
	 1, 2, 3, 4, 5, 6, 7, 8, 9,10,11,12,13,14,15,16,
	17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,
};

static struct ordering KeyTr1 = {
	57,49,41,33,25,17, 9, 1,58,50,42,34,26,18,
	10, 2,59,51,43,35,27,19,11, 3,60,52,44,36,
	63,55,47,39,31,23,15, 7,62,54,46,38,30,22,
	14, 6,61,53,45,37,29,21,13, 5,28,20,12, 4,
};

struct ordering KeyTr2 = {
	14,17,11,24, 1, 5, 3,28,15, 6,21,10,
	23,19,12, 4,26, 8,16, 7,27,20,13, 2,
	41,52,31,37,47,55,30,40,51,45,33,48,
	44,49,39,56,34,53,46,42,50,36,29,32,
};

static struct ordering etr = {
	32, 1, 2, 3, 4, 5, 4, 5, 6, 7, 8, 9,
	 8, 9,10,11,12,13,12,13,14,15,16,17,
	16,17,18,19,20,21,20,21,22,23,24,25,
	24,25,26,27,28,29,28,29,30,31,32, 1,
};

struct ordering ptr = {
	16, 7,20,21,29,12,28,17, 1,15,23,26, 5,18,31,10,
	 2, 8,24,14,32,27, 3, 9,19,13,30, 6,22,11, 4,25,
};

char s_boxes[8][64] = {
{	14, 4,13, 1, 2,15,11, 8, 3,10, 6,12, 5, 9, 0, 7,
	 0,15, 7, 4,14, 2,13, 1,10, 6,12,11, 9, 5, 3, 8,
	 4, 1,14, 8,13, 6, 2,11,15,12, 9, 7, 3,10, 5, 0,
	15,12, 8, 2, 4, 9, 1, 7, 5,11, 3,14,10, 0, 6,13,
},

{	15, 1, 8,14, 6,11, 3, 4, 9, 7, 2,13,12, 0, 5,10,
	 3,13, 4, 7,15, 2, 8,14,12, 0, 1,10, 6, 9,11, 5,
	 0,14, 7,11,10, 4,13, 1, 5, 8,12, 6, 9, 3, 2,15,
	13, 8,10, 1, 3,15, 4, 2,11, 6, 7,12, 0, 5,14, 9,
},

{	10, 0, 9,14, 6, 3,15, 5, 1,13,12, 7,11, 4, 2, 8,
	13, 7, 0, 9, 3, 4, 6,10, 2, 8, 5,14,12,11,15, 1,
	13, 6, 4, 9, 8,15, 3, 0,11, 1, 2,12, 5,10,14, 7,
	 1,10,13, 0, 6, 9, 8, 7, 4,15,14, 3,11, 5, 2,12,
},

{	 7,13,14, 3, 0, 6, 9,10, 1, 2, 8, 5,11,12, 4,15,
	13, 8,11, 5, 6,15, 0, 3, 4, 7, 2,12, 1,10,14, 9,
	10, 6, 9, 0,12,11, 7,13,15, 1, 3,14, 5, 2, 8, 4,
	 3,15, 0, 6,10, 1,13, 8, 9, 4, 5,11,12, 7, 2,14,
},

{	 2,12, 4, 1, 7,10,11, 6, 8, 5, 3,15,13, 0,14, 9,
	14,11, 2,12, 4, 7,13, 1, 5, 0,15,10, 3, 9, 8, 6,
	 4, 2, 1,11,10,13, 7, 8,15, 9,12, 5, 6, 3, 0,14,
	11, 8,12, 7, 1,14, 2,13, 6,15, 0, 9,10, 4, 5, 3,
},

{	12, 1,10,15, 9, 2, 6, 8, 0,13, 3, 4,14, 7, 5,11,
	10,15, 4, 2, 7,12, 9, 5, 6, 1,13,14, 0,11, 3, 8,
	 9,14,15, 5, 2, 8,12, 3, 7, 0, 4,10, 1,13,11, 6,
	 4, 3, 2,12, 9, 5,15,10,11,14, 1, 7, 6, 0, 8,13,
},

{	 4,11, 2,14,15, 0, 8,13, 3,12, 9, 7, 5,10, 6, 1,
	13, 0,11, 7, 4, 9, 1,10,14, 3, 5,12, 2,15, 8, 6,
	 1, 4,11,13,12, 3, 7,14,10,15, 6, 8, 0, 5, 9, 2,
	 6,11,13, 8, 1, 4,10, 7, 9, 5, 0,15,14, 2, 3,12,
},

{	13, 2, 8, 4, 6,15,11, 1,10, 9, 3,14, 5, 0,12, 7,
	 1,15,13, 8,10, 3, 7, 4,12, 5, 6,11, 0,14, 9, 2,
	 7,11, 4, 1, 9,12,14, 2, 0, 6,10,13,15, 3, 5, 8,
	 2, 1,14, 7, 4,10, 8,13,15,12, 9, 0, 3, 5, 6,11,
},
};

int rots[] = {
	1,1,2,2,2,2,2,2,1,2,2,2,2,2,2,1,
};

/* transpose, rotate, and f were rewritten in assembly- see crypta.asm */
#ifndef NOTDEFINED
void __crypt_f __P((int i, struct block *key, struct block *a, struct block *x));
void __crypt_transpose __P((struct block *data, struct ordering *t, int n));
void __crypt_rotate __P((struct block *key));
#else

static
transpose(data, t, n)
	register struct block *data;
	register struct ordering *t;
	register int n;
{
	struct block x;

        memcpy(&x,data,sizeof(struct block));
	/* x = *data; you can't do that in C!!! */

	while (n-- > 0) {
		data->b_data[n] = x.b_data[t->o_data[n] - 1];
	}
}

static
rotate(key)
	register struct block *key;
{
	register unsigned char *p = key->b_data;
	register unsigned char *ep = &(key->b_data[55]);
	int data0 = key->b_data[0], data28 = key->b_data[28];

	while (p++ < ep) *(p-1) = *p;
	key->b_data[27] = data0;
	key->b_data[55] = data28;
}

#endif

struct ordering *EP = &etr;

#ifdef NOTDEFINED
static
f(i, key, a, x)
	struct block *key, *a;
	register struct block *x;
{
	struct block e, ikey, y;
	int k;
	register unsigned char *p, *q, *r;

	memcpy(&e,a,sizeof(struct block));
	/* e = *a; */
	transpose(&e, EP, 48);
	for (k = rots[i]; k; k--) rotate(key);
	memcpy(&ikey,key,sizeof(struct block));
	/* ikey = *key; */
	transpose(&ikey, &KeyTr2, 48);
	p = &(y.b_data[48]);
	q = &(e.b_data[48]);
	r = &(ikey.b_data[48]);
	while (p > y.b_data) {
		*--p = *--q ^ *--r;
	}
	q = x->b_data;
	for (k = 0; k < 8; k++) {
		register int xb, r;

		r = *p++ << 5;
		r += *p++ << 3;
		r += *p++ << 2;
		r += *p++ << 1;
		r += *p++;
		r += *p++ << 4;

		xb = s_boxes[k][r];

		*q++ = (xb >> 3) & 1;
		*q++ = (xb>>2) & 1;
		*q++ = (xb>>1) & 1;
		*q++ = (xb & 1);
	}
	transpose(x, &ptr, 32);
}
#endif

void
setkey(register const char *k)
{

	key = *((struct block *) k);
	transpose(&key, &KeyTr1, 56);
}

void
encrypt(char *blck, int edflag)
{
	register struct block *p = (struct block *) blck;
	register int i;

	transpose(p, &InitialTr, 64);
	for (i = 15; i>= 0; i--) {
		int j = edflag ? i : 15 - i;
		register int k;
		struct block b, x;

		memcpy(&b,p,sizeof(struct block));
		/* b = *p; */
		for (k = 31; k >= 0; k--) {
			p->b_data[k] = b.b_data[k + 32];
		}
		f(j, &key, p, &x);
		for (k = 31; k >= 0; k--) {
			p->b_data[k+32] = b.b_data[k] ^ x.b_data[k];
		}
	}
	transpose(p, &swap, 64);
	transpose(p, &FinalTr, 64);
}

char *
crypt(register const char *pw, const char *salt)
{
	/*	Unfortunately, I had to look at the sources of V7 crypt.
		There was no other way to find out what this routine
		actually does.
	*/
	
	char pwb[66];
	static char result[16];
	register char *p = pwb;
	struct ordering new_etr;
	register int i;
	char *pw2;

	while (*pw && p < &pwb[64]) {
		register int j = 7;

		while (j--) {
			*p++ = (*pw >> j) & 01;
		}
		pw++;
		*p++ = 0;
	}
	while (p < &pwb[64]) *p++ = 0;

	setkey(p = pwb);

	while (p < &pwb[66]) *p++ = 0;

	new_etr = etr;
	EP = &new_etr;
	for (i = 0; i < 2; i++) {
		register char c = *salt++;
		register int j;

		result[i] = c;
		if ( c > 'Z') c -= 6 + 7 + '.';	/* c was a lower case letter */
		else if ( c > '9') c -= 7 + '.';/* c was upper case letter */
		else c -= '.';			/* c was digit, '.' or '/'. */
						/* now, 0 <= c <= 63 */
		for (j = 0; j < 6; j++) {
			if ((c >> j) & 01) {
				int t = 6*i + j;
				int temp = new_etr.o_data[t];
				new_etr.o_data[t] = new_etr.o_data[t+24];
				new_etr.o_data[t+24] = temp;
			}
		}
	}

	if (result[1] == 0) result[1] = result[0];

	for (i = 0; i < 25; i++) encrypt(pwb,0);
	EP = &etr;

	p = pwb;
	pw2 = result+2;
	while (p < &pwb[66]) {
		register int c = 0;
		register int j = 6;

		while (j--) {
			c <<= 1;
			c |= *p++;
		}
		c += '.';		/* becomes >= '.' */
		if (c > '9') c += 7;	/* not in [./0-9], becomes upper */
		if (c > 'Z') c += 6;	/* not in [A-Z], becomes lower */
		*pw2++ = c;
	}
	*pw2 = 0;
	return result;
}
